impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java

```diff
@@ -271,30 +298,40 @@ public void put(final K key, final V value) throws CacheAccessException {
 
     final AtomicBoolean entryActuallyAdded = new AtomicBoolean();
     final long now = timeSource.getTimeMillis();
-
+    
+    final AtomicLong replacedValueSize = new AtomicLong(0l);
     try {
       OnHeapValueHolder<V> valuePut = map.compute(key, new BiFunction<K, OnHeapValueHolder<V>, OnHeapValueHolder<V>>() {
         @Override
         public OnHeapValueHolder<V> apply(K mappedKey, OnHeapValueHolder<V> mappedValue) {
-          entryActuallyAdded.set(mappedValue == null);
 
           if (mappedValue != null && mappedValue.isExpired(now, TimeUnit.MILLISECONDS)) {
             mappedValue = null;
           }
 
+          OnHeapValueHolder<V> toReturn = null;
           if (mappedValue == null) {
-            return newCreateValueHolder(key, value, now);
+            toReturn = newCreateValueHolder(key, value, now);
           } else {
-            return newUpdateValueHolder(key, mappedValue, value, now);
+            replacedValueSize.set(sizeOfEngine.sizeof(mappedValue.value()));// get it only for mapped value
+            toReturn = newUpdateValueHolder(key, mappedValue, value, now);
           }
+          entryActuallyAdded.set(toReturn != null);
+          return toReturn;
         }
       });
 
       if (entryActuallyAdded.get()) {
         putObserver.end(StoreOperationOutcomes.PutOutcome.PUT);
-        enforceCapacity(1);
+        long delta = 1;
+        if (byteSized) { 
+          delta = valuePut.size();
+        }
+        enforceCapacity(delta);
       } else {
         putObserver.end(StoreOperationOutcomes.PutOutcome.REPLACED);
+        long replacedDelta = sizeOfEngine.sizeof(value) - replacedValueSize.get();
```

Comment:
Between the call to sizeof here and the one above on line 316, sizing of a single value happens multiple times. That does not seem like a good thing to do.